### PR TITLE
[Feature] Support ON Clause subquery

### LINKED ISSUE #9875: [Feature] Support ON clause subquery
## Feature request

Currently, we have support subquery in `WHERE Clause`, `HAVING Clause` AND `SELECT Clause`, and we are planning to support subquery in `ON Clause`, examples shown as below:

```sql
-- Un-correlated scalar agg subquery
SELECT * FROM
t0 JOIN t1
ON t0.v1 = t1.v1
AND t0.v2 = (SELECT MAX(v2) FROM t2);

-- Un-correlated scalar non-agg subquery
SELECT * FROM
t0 JOIN t1
ON t0.v1 = t1.v1
AND t0.v2 = (SELECT v2 FROM t2 WHERE v3 < 5);

-- Correlated scalar agg subquery
SELECT * FROM
t0 JOIN t1
ON t0.v1 = t1.v1
AND t0.v2 = (SELECT MAX(v2) FROM t2 WHERE t2.v1 = t0.v1);

-- Un-correlated scalar non-agg subquery
SELECT * FROM
t0 JOIN t1
ON t0.v1 = t1.v1
AND t0.v2 = (SELECT v2 FROM t2 WHERE t2.v1 = t0.v1);

-- Correlated exists subquery
SELECT * FROM
t0 JOIN t1
ON t0.v1 = t1.v1
AND t0.v2 = (exists (SELECT MAX(v2) FROM t2 WHERE t2.v1 = t0.v1));

-- Un-correlated exists subquery
SELECT * FROM
t0 JOIN t1
ON t0.v1 = t1.v1
AND t0.v2 = (exists (SELECT MAX(v2) FROM t2));

-- Correlated quantified subquery
SELECT * FROM
t0 JOIN t1
ON t0.v1 = t1.v1
AND t0.v2 in (SELECT v2 FROM t2 WHERE t2.v1 = t0.v1);

-- Un-correlated quantified subquery
SELECT * FROM
t0 JOIN t1
ON t0.v1 = t1.v1
AND t0.v2 in (SELECT MAX(v2) FROM t2);
```

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #9875

## Enhancement

Since the exist sql parser supports the grammar of ON clause subquery, the only thing we need to do is removing the semantic checking and transforming ON Clause subquery into ApplyOperator

Below are the supported types of ON Clause subquery

-  ✅ orrelated scalar agg subquery
-  ✅ correlated scalar non-agg subquery
-  ✅ uncorrelated scalar agg subquery
-  ✅ uncorrelated scalar non-agg subquery
-  ✅ correlated exist subquery
-  ✅ correlated quantified subquery
-  ✅ uncorrelated exist subquery
-  ✅ uncorrelated quantified subquery

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/Relation.java [hunk 1]
```diff
@@ -46,4 +49,37 @@ public TableName getResolveTableName() {
     public <R, C> R accept(AstVisitor<R, C> visitor, C context) {
         return visitor.visitRelation(this, context);
     }
+
+    public static <T extends Relation> List<T> extractRelations(Relation relation, Class<T> clazz) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/Relation.java:53]: Unused？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java [hunk 2]
```diff
@@ -663,4 +673,82 @@ public LogicalPlan visitTableFunction(TableFunctionRelation node, ExpressionMapp
                 new ExpressionMapping(new Scope(RelationId.of(node), node.getRelationFields()), outputColumns)),
                 null, null);
     }
+
+    private Pair<OptExprBuilder, OptExprBuilder> processOnClauseSubquery(JoinRelation node, OptExprBuilder leftOpt,
+                                                                         OptExprBuilder rightOpt,
+                                                                         ExpressionMapping expressionMapping) {
+        Subquery subquery = node.getOnPredicate().getSubquery();
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java:680]: The method only support one CTE? May this be a problem?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java [hunk 3]
```diff
@@ -663,4 +673,82 @@ public LogicalPlan visitTableFunction(TableFunctionRelation node, ExpressionMapp
...
+        }
+        QueryStatement subqueryStmt = subquery.getQueryStatement();
+        SelectRelation selectRelation = (SelectRelation) subqueryStmt.getQueryRelation();
+        RelationId subqueryRelationId = selectRelation.getRelation().getScope().getRelationId();
+        List<FieldId> correlatedFieldIds = selectRelation.getColumnReferences().values().stream()
+                .filter(field -> !Objects.equals(subqueryRelationId, field.getRelationId()))
+                .collect(Collectors.toList());
+
+        final boolean isLeft;
+        SubqueryTransformer subqueryTransformer = new SubqueryTransformer(session);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java:692]: add more comment plz

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/SubqueryTest.java [hunk 4]
```diff
@@ -569,4 +569,285 @@ public void testCorrelatedScalarNonAggSubqueryWithExpression() throws Exception
...
+                    "  |  colocate: false, reason: \n" +
+                    "  |  equal join conjunct: 1: v1 = 4: v4");
+            assertContains(plan, "  4:HASH JOIN\n" +
+                    "  |  join op: LEFT OUTER JOIN (BROADCAST)\n" +
+                    "  |  colocate: false, reason: \n" +
+                    "  |  equal join conjunct: 2: v2 = 8: v5");
+        }
+        {
+            String sql = "select * from t0 " +
+                    "join t1 on t0.v1 = (select v4 from t1 where t0.v2 = t1.v5 and t1.v5 = 10 and t1.v6 != 3)";
```
Comment (Seaven) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/SubqueryTest.java:688]: can be support complex expression? like: `join t1 on t0.v1 = (t1.v2 in (xxxxxx))` or `join t1 on t0.v1 = (exists (xxxxxx))`?

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/Relation.java (hunk 1 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/Relation.java:53]: Yes, I'll remove it

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java (hunk 2 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java:680]: Yes, it is. Currently, predicate can only has at most one subquery, and I think this limitation canbe removed. Should I handle this in this pr or another pr?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java (hunk 3 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java:692]: Ok

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/SubqueryTest.java (hunk 4 above)
Comment (liuyehcf) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/SubqueryTest.java:688]: Yes, we supporte this kind of subquery

**Comment by liuyehcf** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java [hunk 5]
```diff
@@ -663,4 +673,105 @@ public LogicalPlan visitTableFunction(TableFunctionRelation node, ExpressionMapp
...
+         * TODO, kind == BOTH, both of the left and right relations should be taken into account
+         */
+        final int LEFT = 1;
+        final int RIGHT = 2;
+        final int BOTH = 3;
+        final int kind;
+        SubqueryTransformer subqueryTransformer = new SubqueryTransformer(session);
+        if (correlatedFieldIds.isEmpty()) {
+            // Find the smallest part of join on predicate that contains the subquery
+            Expr subqueryConjunct = node.getOnPredicate();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java:709]: Abstract  this logic to a special method.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java:680]: fine, you can test two subquery in on-cluase later

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Review by Seaven — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 137 / 148 (92.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|2|4|50.00%|[252, 253]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|26|32|81.25%|[378, 393, 394, 406, 440, 456]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SubqueryTransformer.java|24|27|88.89%|[109, 123, 133]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|79|79|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/ExpressionMapping.java|4|4|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java (hunk 5 above)
Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java:725]: I also aggree abstract this logic to a special method. May be subquery related class

Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java]: Done

**Review by Seaven — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SubqueryTransformer.java [hunk 6]
```diff
@@ -104,6 +104,43 @@ public ScalarOperator rewriteSubqueryScalarOperator(Expr predicate, OptExprBuild
         return scalarPredicate;
     }
 
+    public Expr rewriteJoinOnPredicate(Expr predicate) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SubqueryTransformer.java:107]: It looks like a lot of repetitive code, maybe you need mark @todo to refactor their ~

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SubqueryTransformer.java:107]: Of course, it's not important.

**Review by kangkaisen — APPROVED** [review]

**Comment by kangkaisen** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 140 / 150 (93.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|2|4|50.00%|[252, 253]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|27|32|84.38%|[383, 398, 399, 411, 461]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SubqueryTransformer.java|24|27|88.89%|[109, 123, 133]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|80|80|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/LoadScanNode.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/ExpressionMapping.java|4|4|100.00%|[]|

**Pull request merged**